File: remix-of-vesper-blog-main/src/services/commitmentService.ts
// Commitment service — buyer→farmer commitment workflow
// Uses client-side Firestore (same pattern as demands)
import {
    collection,
    doc,
    addDoc,
    getDoc,
    getDocs,
    updateDoc,
    query,
    where,
    serverTimestamp,
    increment,
} from "firebase/firestore";
import { db, auth } from "@/firebase";
import { createShipmentFromCommitment } from "@/services/shipmentService";

// -------- Types --------

export type CommitmentStatus = "pending" | "accepted" | "rejected" | "completed";

export interface CommitmentDocument {
    id: string;
    cropId: string;
    cropName: string;
    itemType: "primary" | "byproduct";
    itemName: string;
    farmerId: string;
    buyerId: string;
    agreedPrice: number;
    quantity: number;
    status: CommitmentStatus;
    createdAt: Date | null;
    updatedAt: Date | null;
}

export interface CreateCommitmentPayload {
    cropId: string;
    agreedPrice: number;
    quantity: number;
    itemType?: "primary" | "byproduct";
    itemName?: string;
}

// Valid status transitions
const VALID_TRANSITIONS: Record<string, string[]> = {
    pending: ["accepted", "rejected"],
    accepted: ["completed"],
    rejected: [],
    completed: [],
};

// -------- Helpers --------

function getUid(): string | null {
    let uid = auth.currentUser?.uid;
    if (!uid) {
        try {
            const s = sessionStorage.getItem("vasudha_user");
            if (s) uid = JSON.parse(s).id;
        } catch { /* ignore */ }
    }
    return uid || null;
}

// -------- Create Commitment --------

/**
 * Create a commitment from buyer to farmer.
 * Validates: crop exists, is active, quantity available.
 */
export async function createCommitment(payload: CreateCommitmentPayload): Promise<string> {
    const buyerId = getUid();
    if (!buyerId) throw new Error("Not authenticated");

    // 1. Validate crop exists and is active
    const cropRef = doc(db, "crops", payload.cropId);
    const cropSnap = await getDoc(cropRef);

    if (!cropSnap.exists()) {
        throw new Error("Crop not found");
    }

    const crop = cropSnap.data();

    if (crop.status !== "active" && crop.status !== "partially_committed") {
        throw new Error("Crop is not available for commitment");
    }

    const itemType = payload.itemType || "primary";
    const itemName = payload.itemName || "";

    // 2. Validate quantity based on item type
    if (itemType === "byproduct") {
        const byproducts: any[] = crop.byproducts || [];
        const bpMatch = byproducts.find((bp: any) => bp.name === itemName);
        if (!bpMatch) {
            throw new Error(`By-product '${itemName}' not found for this crop`);
        }
        const bpAvailable = (bpMatch.availableQuantity ?? bpMatch.totalQuantity ?? 0) - (bpMatch.committedQuantity || 0);
        if (payload.quantity <= 0) throw new Error("Quantity must be greater than 0");
        if (payload.quantity > bpAvailable) {
            throw new Error(`Only ${bpAvailable}q of '${itemName}' available`);
        }
    } else {
        // Primary crop
        const estimatedYield = crop.estimatedYield || 0;
        const committedQty = crop.committedQuantity || 0;
        const availableQty = estimatedYield - committedQty;

        if (payload.quantity <= 0) {
            throw new Error("Quantity must be greater than 0");
        }

        if (payload.quantity > availableQty) {
            throw new Error(`Only ${availableQty}q available (${committedQty}q already committed)`);
        }
    }

    // 3. Create commitment document
    const commitmentData = {
        cropId: payload.cropId,
        cropName: crop.cropName || "",
        itemType: itemType,
        itemName: itemType === "byproduct" ? itemName : (crop.cropName || ""),
        farmerId: crop.farmerId || "",
        buyerId: buyerId,
        agreedPrice: payload.agreedPrice,
        quantity: payload.quantity,
        status: "pending" as CommitmentStatus,
        createdAt: serverTimestamp(),
        updatedAt: serverTimestamp(),
    };

    const docRef = await addDoc(collection(db, "commitments"), commitmentData);
    console.log("[CommitmentService] Created commitment:", docRef.id);
    return docRef.id;
}

// -------- Update Status --------

/**
 * Update commitment status with controlled transitions.
 * On accept → updates crop's committedQuantity.
 */
export async function updateCommitmentStatus(
    commitmentId: string,
    newStatus: CommitmentStatus
): Promise<void> {
    const uid = getUid();
    if (!uid) throw new Error("Not authenticated");

    // 1. Get current commitment
    const commitRef = doc(db, "commitments", commitmentId);
    const commitSnap = await getDoc(commitRef);

    if (!commitSnap.exists()) {
        throw new Error("Commitment not found");
    }

    const commitment = commitSnap.data();
    const currentStatus = commitment.status as CommitmentStatus;

    // 2. Validate transition
    const allowed = VALID_TRANSITIONS[currentStatus] || [];
    if (!allowed.includes(newStatus)) {
        throw new Error(`Cannot transition from "${currentStatus}" to "${newStatus}"`);
    }

    // 3. Validate permissions
    if ((newStatus === "accepted" || newStatus === "rejected") && uid !== commitment.farmerId) {
        throw new Error("Only the farmer can accept or reject commitments");
    }

    // 4. Update commitment
    await updateDoc(commitRef, {
        status: newStatus,
        updatedAt: serverTimestamp(),
    });

    // 5. If accepted → update crop's committed quantity
    if (newStatus === "accepted") {
        const cropRef = doc(db, "crops", commitment.cropId);
        const cropSnap = await getDoc(cropRef);

        if (cropSnap.exists()) {
            const crop = cropSnap.data();
            const newCommitted = (crop.committedQuantity || 0) + (commitment.quantity || 0);
            const estimatedYield = crop.estimatedYield || 0;

            const cropUpdate: Record<string, any> = {
                committedQuantity: increment(commitment.quantity || 0),
                updatedAt: serverTimestamp(),
            };

            // If fully committed, update status
            if (newCommitted >= estimatedYield) {
                cropUpdate.status = "fully_committed";
            } else if (newCommitted > 0) {
                cropUpdate.status = "partially_committed";
            }

            await updateDoc(cropRef, cropUpdate);
            console.log("[CommitmentService] Updated crop committed quantity:", newCommitted);
        }

        // 6. Auto-create shipment
        try {
            const shipmentId = await createShipmentFromCommitment({
                id: commitmentId,
                cropId: commitment.cropId,
                cropName: commitment.cropName || "",
                farmerId: commitment.farmerId,
                buyerId: commitment.buyerId,
                quantity: commitment.quantity || 0,
            });
            console.log("[CommitmentService] Auto-created shipment:", shipmentId);
        } catch (shipErr) {
            console.error("[CommitmentService] Shipment auto-creation failed (non-blocking):", shipErr);
        }
    }

    console.log("[CommitmentService] Updated commitment", commitmentId, "→", newStatus);
}

// -------- Query Commitments --------

/**
 * Get all commitments for a buyer.
 */
export async function getCommitmentsForBuyer(): Promise<CommitmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];

    try {
        const q = query(
            collection(db, "commitments"),
            where("buyerId", "==", uid)
        );
        const snap = await getDocs(q);
        return snap.docs.map((d) => {
            const data = d.data();
            return {
                id: d.id,
                cropId: data.cropId || "",
                cropName: data.cropName || "",
                itemType: data.itemType || "primary",
                itemName: data.itemName || data.cropName || "",
                farmerId: data.farmerId || "",
                buyerId: data.buyerId || "",
                agreedPrice: data.agreedPrice || 0,
                quantity: data.quantity || 0,
                status: (data.status || "pending") as CommitmentStatus,
                createdAt: data.createdAt?.toDate?.() || null,
                updatedAt: data.updatedAt?.toDate?.() || null,
            };
        });
    } catch (err) {
        console.error("[CommitmentService] getCommitmentsForBuyer failed:", err);
        return [];
    }
}

/**
 * Get all commitments for a farmer.
 */
export async function getCommitmentsForFarmer(): Promise<CommitmentDocument[]> {
    const uid = getUid();
    if (!uid) return [];

    try {
        const q = query(
            collection(db, "commitments"),
            where("farmerId", "==", uid)
        );
        const snap = await getDocs(q);
        return snap.docs.map((d) => {
            const data = d.data();
            return {
                id: d.id,
                cropId: data.cropId || "",
                cropName: data.cropName || "",
                itemType: data.itemType || "primary",
                itemName: data.itemName || data.cropName || "",
                farmerId: data.farmerId || "",
                buyerId: data.buyerId || "",
                agreedPrice: data.agreedPrice || 0,
                quantity: data.quantity || 0,
                status: (data.status || "pending") as CommitmentStatus,
                createdAt: data.createdAt?.toDate?.() || null,
                updatedAt: data.updatedAt?.toDate?.() || null,
            };
        });
    } catch (err) {
        console.error("[CommitmentService] getCommitmentsForFarmer failed:", err);
        return [];
    }
}

File: backend/app/services/commitment_service.py
"""
Commitment service — buyer→farmer commitment workflow (backend version).
Supports both primary crop and by-product commitments.
"""


def create_commitment(db, data: dict) -> str:
    """Create a new commitment. Validates crop and quantity.
    
    data must include:
        cropId, buyerId, quantity, agreedPrice
        itemType: "primary" or "byproduct" (default: "primary")
        itemName: name of by-product (required if itemType == "byproduct")
    """
    crop_ref = db.collection("crops").document(data["cropId"])
    crop_snap = crop_ref.get()

    if not crop_snap.exists:
        raise ValueError("Crop not found")

    crop = crop_snap.to_dict()

    if crop.get("status") not in ("active", "partially_committed"):
        raise ValueError("Crop is not available for commitment")

    item_type = data.get("itemType", "primary")
    item_name = data.get("itemName", "")

    if data.get("quantity", 0) <= 0:
        raise ValueError("Quantity must be greater than 0")

    # Validate quantity based on item type
    if item_type == "byproduct":
        # Find the by-product in crop's byproducts array
        byproducts = crop.get("byproducts", [])
        bp_match = None
        bp_index = -1
        for i, bp in enumerate(byproducts):
            if bp.get("name") == item_name:
                bp_match = bp
                bp_index = i
                break

        if bp_match is None:
            raise ValueError(f"By-product '{item_name}' not found for this crop")

        available_qty = bp_match.get("availableQuantity", bp_match.get("totalQuantity", 0)) - bp_match.get("committedQuantity", 0)

        if data["quantity"] > available_qty:
            raise ValueError(f"Only {available_qty}q of '{item_name}' available ({bp_match.get('committedQuantity', 0)}q already committed)")
    else:
        # Primary crop
        estimated_yield = crop.get("estimatedYield", 0)
        committed_qty = crop.get("committedQuantity", 0)
        available_qty = estimated_yield - committed_qty

        if data["quantity"] > available_qty:
            raise ValueError(f"Only {available_qty}q available ({committed_qty}q already committed)")

    from datetime import datetime, timezone

    commitment_data = {
        "cropId": data["cropId"],
        "cropName": crop.get("cropName", ""),
        "farmerId": crop.get("farmerId", ""),
        "buyerId": data["buyerId"],
        "agreedPrice": float(data.get("agreedPrice", 0)),
        "quantity": float(data["quantity"]),
        "itemType": item_type,
        "itemName": item_name if item_type == "byproduct" else crop.get("cropName", ""),
        "status": "pending",
        "createdAt": datetime.now(timezone.utc).isoformat(),
        "updatedAt": datetime.now(timezone.utc).isoformat(),
    }

    _, doc_ref = db.collection("commitments").add(commitment_data)
    return doc_ref.id


def update_commitment_status(db, commitment_id: str, new_status: str, user_id: str) -> dict:
    """Update commitment status with controlled transitions."""
    valid_transitions = {
        "pending": ["accepted", "rejected"],
        "accepted": ["completed"],
        "rejected": [],
        "completed": [],
    }

    doc_ref = db.collection("commitments").document(commitment_id)
    snap = doc_ref.get()

    if not snap.exists:
        raise ValueError("Commitment not found")

    commitment = snap.to_dict()
    current = commitment.get("status", "")

    allowed = valid_transitions.get(current, [])
    if new_status not in allowed:
        raise ValueError(f"Cannot transition from '{current}' to '{new_status}'")

    if new_status in ("accepted", "rejected") and user_id != commitment.get("farmerId"):
        raise ValueError("Only the farmer can accept or reject commitments")

    from datetime import datetime, timezone
    from google.cloud.firestore import Increment

    doc_ref.update({
        "status": new_status,
        "updatedAt": datetime.now(timezone.utc).isoformat(),
    })

    # On accept → update inventory (main crop or by-product)
    if new_status == "accepted":
        crop_ref = db.collection("crops").document(commitment["cropId"])
        crop_snap = crop_ref.get()
        if crop_snap.exists:
            crop = crop_snap.to_dict()
            qty = commitment.get("quantity", 0)
            item_type = commitment.get("itemType", "primary")
            item_name = commitment.get("itemName", "")

            if item_type == "byproduct":
                # Update by-product committed quantity
                byproducts = crop.get("byproducts", [])
                for i, bp in enumerate(byproducts):
                    if bp.get("name") == item_name:
                        byproducts[i]["committedQuantity"] = bp.get("committedQuantity", 0) + qty
                        byproducts[i]["availableQuantity"] = bp.get("totalQuantity", 0) - byproducts[i]["committedQuantity"]
                        break
                crop_ref.update({
                    "byproducts": byproducts,
                    "updatedAt": datetime.now(timezone.utc).isoformat(),
                })
            else:
                # Update main crop committed quantity
                new_committed = crop.get("committedQuantity", 0) + qty
                estimated = crop.get("estimatedYield", 0)

                crop_update = {
                    "committedQuantity": Increment(qty),
                    "updatedAt": datetime.now(timezone.utc).isoformat(),
                }
                if new_committed >= estimated:
                    crop_update["status"] = "fully_committed"
                elif new_committed > 0:
                    crop_update["status"] = "partially_committed"

                crop_ref.update(crop_update)

    return {"commitmentId": commitment_id, "status": new_status}


def get_commitments_for_farmer(db, farmer_id: str) -> list[dict]:
    """Get all commitments for a farmer."""
    docs = db.collection("commitments").where("farmerId", "==", farmer_id).stream()
    return [{"id": d.id, **d.to_dict()} for d in docs]


def get_commitments_for_buyer(db, buyer_id: str) -> list[dict]:
    """Get all commitments for a buyer."""
    docs = db.collection("commitments").where("buyerId", "==", buyer_id).stream()
    return [{"id": d.id, **d.to_dict()} for d in docs]
